crates/js-sys/src/lib.rs

```diff
@@ -1538,12 +1471,37 @@ pub fn try_iter(val: &JsValue) -> Result<Option<IntoIter>, JsValue> {
 
     Ok(if next.is_function() {
         let it: Iterator = it.unchecked_into();
-        Some(it.into_iter())
+        Some(it)
     } else {
         None
     })
 }
 
+/// An `Iterator` wrapper for when we know the type that will be returned.
+///
+/// We also assume that errors from the use of this iterator are not possible.
+pub struct TypedIterator<T> {
```

Comment:
I do like the idea of building typed iterators on top, though.

Unsure if it belongs in js-sys or a third party crate.

Maybe want to make `Item = Result<T>` to handle potential exceptions and also use checked casts that return `Err` on cast failures?

Reply:
The primary purpose of the typed iterator is to provide for iteration on web-sys types.

For example, `NodeList` is an iterator of `Node`. And so we want to provide a `TypedIterator<Node>` for `NodeList`.

Since we know that the `NodeList` iterator doesn't error, we don't want it to return `Result`.

That's why it uses `JsCast::unchecked_from_js`, and why it's called `TypedIterator::from_iterator_unchecked`, to emphasize the unchecked nature of it.

I think it should be in js-sys, because:

1) web-sys needs it, and...

2) ...a lot of JS types outside of web-sys also define iterators, so we should have a convenient zero-cost way for users to create iterators for those types.

3) It fits in nicely with the existing iterator types.

Reply:
>     1. web-sys needs it, and...
> 
>     2. ...a lot of JS types outside of web-sys also define iterators, so we should have a convenient zero-cost way for users to create iterators for those types.

I think these two use cases call for different iterator types, though. For web-sys we can reasonably trust that the iterator will yield items that it says it will, and therefore it makes sense to use unchecked casts. For (2) I don't think that is the case at all, and we should be using checked casts by default.

Reply:
> For (2) I don't think that is the case at all, and we should be using checked casts by default.

I disagree. When somebody creates bindings for an existing JS iterator, they know whether the JS iterator will throw or not.

Requiring all users of the iterator to `unwrap` the `Result` is really bad for ergonomics, and it doesn't give any extra safety either (since they know that the iterator cannot error).

It's not any different from how `JsCast` provides both `unchecked_into` and `dyn_into`. This is the `unchecked_into` version of iterators.

There is prior precedent for it as well: the `wasm_bindgen` macro by default does *not* catch errors, you have to specifically use `catch` to enable catching errors.

Having a checked iterator type sounds reasonable, but there should definitely be an unchecked iterator type as well (which is what this type is).

So are you proposing to have both a checked and unchecked typed iterator? Because I would be completely fine with that.

Reply:
> So are you proposing to have both checked _and_ unchecked typed iterators? Because I would be completely fine with that.

Yes.